src/Poseidon/EntitiesList.hs

```diff
@@ -39,63 +40,98 @@ instance Show SignedEntity where
 
-isInclude :: SignedEntity -> Bool
-isInclude (Include _) = True
-isInclude _ = False
-
-isExclude :: SignedEntity -> Bool
-isExclude (Exclude _) = True
-isExclude _ = False
-
-fromSignedEntity :: SignedEntity -> PoseidonEntity
-fromSignedEntity (Include a) = a
-fromSignedEntity (Exclude a) = a
-
-entityIncludes :: SignedEntitiesList -> EntitiesList
-entityIncludes xs = map fromSignedEntity $ filter isInclude xs
-
-entityExcludes :: SignedEntitiesList -> EntitiesList
-entityExcludes xs = map fromSignedEntity $ filter isExclude xs
-
--- | A parser to parse entities
-parseSignedEntitiesList :: P.Parser SignedEntitiesList
-parseSignedEntitiesList = do 
-    eL <- P.sepBy parsePoseidonEntity (P.char ',' <* P.spaces)
-    P.optional (P.many (P.char ' ') >> comment)
-    return eL
-
-parseSignedEntitiesListFromFile :: P.Parser SignedEntitiesList
-parseSignedEntitiesListFromFile = do
-    concat <$> P.sepBy (P.try emptyLine <|> comment <|> parseSignedEntitiesList) P.newline
+-- A class to generalise signed and unsigned Entity Lists. Both have the feature that they can be used to filter individuals.
+class EntitySpec a where
+    indInfoConformsToEntitySpec :: [a] -> IndividualInfo -> Bool
+    underlyingEntity :: a -> PoseidonEntity
+    entitySpecParser :: P.Parser a
+
+instance EntitySpec SignedEntity where
+    indInfoConformsToEntitySpec signedEntities (IndividualInfo indName groupNames pacName) = go signedEntities False
+      where
+        go [] r = r
+        go (Include (Ind   n):rest) r = if n ==     indName    then go rest True  else go rest r
+        go (Include (Group n):rest) r = if n `elem` groupNames then go rest True  else go rest r
+        go (Include (Pac   n):rest) r = if n ==     pacName    then go rest True  else go rest r
+        go (Exclude (Ind   n):rest) r = if n ==     indName    then go rest False else go rest r
+        go (Exclude (Group n):rest) r = if n `elem` groupNames then go rest False else go rest r
+        go (Exclude (Pac   n):rest) r = if n ==     pacName    then go rest False else go rest r
+    underlyingEntity = removeEntitySign
+    entitySpecParser = parseSign <*> entitySpecParser
+      where
+        parseSign = (P.char '-' >> return Exclude) <|> (P.optional (P.char '+') >> return Include)
+
+instance EntitySpec PoseidonEntity where
+    indInfoConformsToEntitySpec entities = indInfoConformsToEntitySpec (map Include entities)
+    underlyingEntity = id
+    entitySpecParser = parsePac <|> parseGroup <|> parseInd
+      where
+        parsePac   = Pac   <$> P.between (P.char '*') (P.char '*') parseName
+        parseGroup = Group <$> parseName
+        parseInd   = Ind   <$> P.between (P.char '<') (P.char '>') parseName
+        parseName  = P.many1 (P.satisfy (\c -> not (isSpace c || c `elem` ",<>*")))
+
+removeEntitySign :: SignedEntity -> PoseidonEntity
+removeEntitySign (Include e) = e
+removeEntitySign (Exclude e) = e
+
+entitiesListP :: (EntitySpec a) => P.Parser [a]
```

Comment:
Yeah, no... I don't think we're very consistent here... I started it in this module, because I got confused previously with the mixed naming. Yes, I'm also a fan of making these things more consistent.

Reply:
At some point we should do some major refactoring to clean up all this heterogeneity. poseidon-hs is probably pretty complete in terms of features. So we could think about cleaning and shrinking the code base.